interfaces/prompting/constraints.go

```diff
@@ -97,17 +129,333 @@ func (c *Constraints) Match(path string) (bool, error) {
 	return match, nil
 }
 
-// ContainPermissions returns true if the constraints include every one of the
-// given permissions.
-func (c *Constraints) ContainPermissions(permissions []string) bool {
-	for _, perm := range permissions {
-		if !strutil.ListContains(c.Permissions, perm) {
+// ReplyConstraints hold information about the applicability of a reply to
+// particular paths or permissions. Upon receiving the reply, snapd converts
+// ReplyConstraints to Constraints.
+type ReplyConstraints struct {
+	PathPattern *patterns.PathPattern `json:"path-pattern"`
+	Permissions []string              `json:"permissions"`
+}
+
+// ToConstraints validates the receiving ReplyConstraints with respect to the
+// given interface, along with the given outcome, lifespan, and duration, and
+// constructs an equivalent Constraints from the ReplyConstraints.
+func (c *ReplyConstraints) ToConstraints(iface string, outcome OutcomeType, lifespan LifespanType, duration string) (*Constraints, error) {
+	if _, err := outcome.AsBool(); err != nil {
+		// Should not occur, as outcome is validated when unmarshalled
+		return nil, err
+	}
+	if _, err := lifespan.ParseDuration(duration, time.Now()); err != nil {
+		return nil, err
+	}
+	if c.PathPattern == nil {
+		return nil, prompting_errors.NewInvalidPathPatternError("", "no path pattern")
+	}
+	availablePerms, ok := interfacePermissionsAvailable[iface]
+	if !ok {
+		return nil, prompting_errors.NewInvalidInterfaceError(iface, availableInterfaces())
+	}
+	if len(c.Permissions) == 0 {
+		return nil, prompting_errors.NewPermissionsListEmptyError(iface, availablePerms)
+	}
+	var invalidPerms []string
+	permissionMap := make(PermissionMap, len(c.Permissions))
+	for _, perm := range c.Permissions {
+		if !strutil.ListContains(availablePerms, perm) {
+			invalidPerms = append(invalidPerms, perm)
+			continue
+		}
+		permissionMap[perm] = &PermissionEntry{
+			Outcome:  outcome,
+			Lifespan: lifespan,
+			Duration: duration,
+		}
+	}
+	if len(invalidPerms) > 0 {
+		return nil, prompting_errors.NewInvalidPermissionsError(iface, invalidPerms, availablePerms)
+	}
+	constraints := &Constraints{
+		PathPattern: c.PathPattern,
+		Permissions: permissionMap,
+	}
+	return constraints, nil
+}
+
+// PatchConstraints hold information about the applicability of the modified
+// rule to particular paths or permissions. A request will be matched by the
+// constraints if the requested path is matched by the path pattern (according
+// to bash's globstar matching) and one or more requested permissions are
+// denied in the permission map, or all of the requested permissions are
+// allowed in the map. When snapd modifies the rule, it converts
+// PatchConstraints to RuleConstraints, using the rule's existing constraints
+// wherever a field is omitted.
```

Comment:
feels like a shorter comment would be less confusing
```suggestion
// PatchConstraints hold information about the applicability of the modified
// rule to particular paths or permissions. When snapd modifies the rule, 
// it converts PatchConstraints to RuleConstraints, using the rule's existing
// constraints wherever a field is omitted.
```